pkg/controller/fakeclusterinstall/fakeclusterinstall_controller.go

```diff
@@ -0,0 +1,313 @@
+type ReconcileClusterInstall struct {
+	client.Client
+	scheme *runtime.Scheme
+	logger log.FieldLogger
+
+	// updateStatus updates a given cluster state's status, exposed for testing
+	updateStatus func(client.Client, *hivev1.FakeClusterInstall, log.FieldLogger) error
+}
+
+// Reconcile ensures that a given FakeClusterInstall resource exists and reflects the state of cluster operators from its target cluster
+func (r *ReconcileClusterInstall) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := controllerutils.BuildControllerLogger(ControllerName, "fakeClusterInstall", request.NamespacedName)
+	logger.Info("reconciling FakeClusterInstall")
+	recobsrv := hivemetrics.NewReconcileObserver(ControllerName, logger)
+	defer recobsrv.ObserveControllerReconcileTime()
+
+	// Fetch the FakeClusterInstall instance
+	fci := &hivev1.FakeClusterInstall{}
+	err := r.Get(context.TODO(), request.NamespacedName, fci)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			logger.Debug("FakeClusterInstall not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.WithError(err).Error("Error getting FakeClusterInstall")
+		return reconcile.Result{}, err
+	}
+
+	if !fci.DeletionTimestamp.IsZero() {
+		logger.Info("FakeClusterInstall resource has been deleted")
+		return reconcile.Result{}, nil
+	}
+
+	// Ensure our conditions are present, default state should be Unknown per Kube guidelines:
+	conditionTypes := []string{
+		// These conditions are required by Hive:
+		hivev1.ClusterInstallCompleted,
+		hivev1.ClusterInstallFailed,
+		hivev1.ClusterInstallStopped,
+		hivev1.ClusterInstallRequirementsMet,
+	}
+
+	var anyChanged bool
+	for _, condType := range conditionTypes {
+		c := controllerutils.FindClusterInstallCondition(fci.Status.Conditions, condType)
+		if c == nil {
+			logger.WithField("condition", condType).Info("initializing condition with Unknown status")
+			newConditions, changed := controllerutils.SetClusterInstallConditionWithChangeCheck(
+				fci.Status.Conditions,
+				condType,
+				corev1.ConditionUnknown,
+				"",
+				"",
+				controllerutils.UpdateConditionAlways)
+			fci.Status.Conditions = newConditions
+			anyChanged = anyChanged || changed
+		}
+	}
+
+	if anyChanged {
+		err := updateClusterInstallStatus(r.Client, fci, logger)
+		return reconcile.Result{}, err
+	}
+
+	// Check if we're already Stopped=True and return if so.
+	// If not, we should be Stopped=False as we're actively working.
+	stoppedCond := controllerutils.FindClusterInstallCondition(fci.Status.Conditions, hivev1.ClusterInstallStopped)
+	if stoppedCond == nil {
+		errMsg := fmt.Sprintf("%s condition is nil but should be set", hivev1.ClusterInstallStopped)
+		return reconcile.Result{}, errors.New(errMsg)
+	}
```

Comment:
personally, i would depend on Installed = true, to ensure that stopped is set to true.

Reply:
That would be technically fine in this controller I just was thinking this might serve as a bit of a reference for anyone implementing one of these. In that case we could still hit Completed=False and Stopped=True. Below I get a little more fakecluster specific and I set Stopped=True when Completed=True. Does that seem reasonable?

Reply:
not reconciling when Stopped = True is a bad recommendation imo. For example, if the impl supported max retries it should also support user increasing the max retries when it previously stopped. This example provides bad advice.

Decide where there is no work to be done on Completed = true or other computable level state.

Reply:
Good point, updating.